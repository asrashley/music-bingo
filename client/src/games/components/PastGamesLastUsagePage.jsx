import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { PastGamesLastUsage } from './PastGamesLastUsage';
import { AdminActions } from '../../admin/components/AdminActions';

import { fetchUserIfNeeded } from '../../user/userSlice';
import { fetchGamesIfNeeded, invalidateGames } from '../gamesSlice';

import { getPastGamesCalendar, isFetchingGames } from '../gamesSelectors';
import { getUser } from '../../user/userSelectors';

import { UserPropType } from '../../user/types/User';

class PastGamesLastUsagePageComponent extends React.Component {
    static propTypes = {
        dispatch: PropTypes.func.isRequired,
        user: UserPropType.isRequired,
        loading: PropTypes.bool.isRequired,
    };

    componentDidMount() {
        const { dispatch } = this.props;
        dispatch(fetchUserIfNeeded());
        dispatch(fetchGamesIfNeeded());
    }

    componentDidUpdate(prevProps, prevState) {
        const { user, dispatch } = this.props;
        if (user.pk > 0 && user.pk !== prevProps.user.pk) {
            dispatch(fetchGamesIfNeeded());
        }
    }

    onReload = () => {
        const { dispatch } = this.props;
        dispatch(invalidateGames());
        dispatch(fetchGamesIfNeeded());
    };


    render() {
        const { pastGamesCalendar, user, loading } = this.props;
        const { themes } = pastGamesCalendar;

        return (
            <div id="games-page" className={user.loggedIn ? '' : 'modal-open'}  >
                <AdminActions />
                <PastGamesLastUsage themes={themes} loading={loading} />
            </div>
        );
    }
}

const mapStateToProps = (state, ownProps) => {
    return {
        loading: isFetchingGames(state, ownProps),
        pastGamesCalendar: getPastGamesCalendar(state, ownProps),
        user: getUser(state, ownProps),
    };
};

export const PastGamesLastUsagePage = connect(mapStateToProps)(PastGamesLastUsagePageComponent);

