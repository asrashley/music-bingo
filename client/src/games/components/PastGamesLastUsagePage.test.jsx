import React from 'react';
import log from 'loglevel';

import { fetchMock, renderWithProviders, installFetchMocks } from '../../../tests';
import { PastGamesLastUsagePage } from './PastGamesLastUsagePage';

describe('PastGamesLastUsagePage component', () => {
    beforeEach(() => {
        installFetchMocks(fetchMock, { loggedIn: true });
    });

    afterEach(() => {
        fetchMock.mockReset();
        log.resetLevel();
    });

    it('to render listing of when each theme was last used', async () => {
        const { asFragment, findAllByText, getByText, store } = renderWithProviders(<PastGamesLastUsagePage />);
        await findAllByText("Rock & Power Ballads", { exact: false });
        const { games } = store.getState().games;
        Object.values(games).forEach(game => {
            getByText(game.title);
        });
        expect(asFragment()).toMatchSnapshot();
    });
});