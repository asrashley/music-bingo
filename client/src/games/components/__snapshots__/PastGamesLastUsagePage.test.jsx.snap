// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PastGamesLastUsagePage component to render listing of when each theme was last used 1`] = `
<DocumentFragment>
  <div
    class=""
    id="games-page"
  >
    <div
      class="action-panel"
    >
      <button
        class="btn btn-primary ml-2"
      >
        Import Game
      </button>
    </div>
    <div
      class="past-games-buttons"
    >
      <div
        class="row mb-4 page-usage"
      >
        <div
          class="col text-center"
        >
          <a
            class="btn btn-lg btn-light border border-dark-subtle popularity-button"
            href="/history"
          >
            <h4>
              Theme Popularity
            </h4>
          </a>
        </div>
        <div
          class="col text-center"
        >
          <a
            class="btn btn-lg btn-light border border-dark-subtle usage-button"
            href="/history/themes"
          >
            <h4>
              Theme Last Usage
            </h4>
          </a>
        </div>
        <div
          class="col text-center"
        >
          <a
            class="btn btn-lg btn-light border border-dark-subtle calendar-button"
            href="/history/calendar"
          >
            <h4>
              Games Calendar
            </h4>
          </a>
        </div>
        <div
          class="col text-center"
        >
          <a
            class="btn btn-lg btn-light border border-dark-subtle all-button"
            href="/history/games"
          >
            <h4>
              All Games
            </h4>
          </a>
        </div>
      </div>
    </div>
    <div
      aria-busy="false"
      aria-colcount="3"
      aria-rowcount="8"
      class="past-games-calendar table table-striped rs-table rs-table-bordered rs-table-hover rs-table-cell-bordered"
      role="grid"
      style="width: 750px; height: 368px;"
      title="Theme History"
    >
      <div
        class="rs-table-header-row-wrapper"
        role="rowgroup"
      >
        <div
          aria-rowindex="1"
          class="rs-table-row rs-table-row-header"
          role="row"
          style="min-width: 750px; height: 40px; transform: translate3d(0px,0px,0);"
        >
          <div
            class="rs-table-cell-group rs-table-cell-group-scroll"
            style="transform: translate3d(0px,0px,0);"
          >
            <div
              class="rs-table-cell-header rs-table-cell-header-sortable"
            >
              <div
                aria-colindex="1"
                class="rs-table-cell rs-table-cell-first"
                role="columnheader"
                style="width: 350px; height: 40px; z-index: 0; left: 0px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 350px; height: 40px; text-align: left;"
                >
                  Theme
                  <span
                    class="rs-table-cell-header-sort-wrapper"
                  >
                    <svg
                      aria-hidden="true"
                      aria-label="sort"
                      class="rs-table-cell-header-icon-sort rs-icon"
                      data-category="direction"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 16 16"
                      width="1em"
                    >
                      <path
                        d="M7.616 3.18a.5.5 0 01.696-.071l.072.071 2.5 3a.5.5 0 01-.704.704l-.064-.064L8 4.281 5.884 6.82a.5.5 0 01-.63.115l-.074-.051a.5.5 0 01-.115-.63l.051-.074 2.5-3zM7.616 12.82a.5.5 0 00.696.071l.072-.071 2.5-3a.5.5 0 00-.704-.704l-.064.064L8 11.719 5.884 9.18a.5.5 0 00-.63-.115l-.074.051a.5.5 0 00-.115.63l.051.074 2.5 3z"
                      />
                    </svg>
                  </span>
                </div>
              </div>
              <div
                class="rs-table-column-resize-spanner"
                role="button"
                style="left: 348px; height: 39px;"
                tabindex="-1"
              />
            </div>
            <div
              class="rs-table-cell-header rs-table-cell-header-sortable"
            >
              <div
                aria-colindex="2"
                class="rs-table-cell"
                role="columnheader"
                style="width: 200px; height: 40px; z-index: 0; left: 350px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 40px; text-align: right;"
                >
                  Last Used
                  <span
                    class="rs-table-cell-header-sort-wrapper"
                  >
                    <svg
                      aria-hidden="true"
                      aria-label="sort"
                      class="rs-table-cell-header-icon-sort rs-icon"
                      data-category="direction"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 16 16"
                      width="1em"
                    >
                      <path
                        d="M7.616 3.18a.5.5 0 01.696-.071l.072.071 2.5 3a.5.5 0 01-.704.704l-.064-.064L8 4.281 5.884 6.82a.5.5 0 01-.63.115l-.074-.051a.5.5 0 01-.115-.63l.051-.074 2.5-3zM7.616 12.82a.5.5 0 00.696.071l.072-.071 2.5-3a.5.5 0 00-.704-.704l-.064.064L8 11.719 5.884 9.18a.5.5 0 00-.63-.115l-.074.051a.5.5 0 00-.115.63l.051.074 2.5 3z"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="rs-table-cell-header rs-table-cell-header-sortable"
            >
              <div
                aria-colindex="3"
                class="rs-table-cell rs-table-cell-last"
                role="columnheader"
                style="width: 200px; height: 40px; z-index: 0; left: 550px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 40px; text-align: right;"
                >
                  Elapsed Time
                  <span
                    class="rs-table-cell-header-sort-wrapper"
                  >
                    <svg
                      aria-hidden="true"
                      aria-label="sort"
                      class="rs-table-cell-header-icon-sort rs-icon"
                      data-category="direction"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 16 16"
                      width="1em"
                    >
                      <path
                        d="M7.616 3.18a.5.5 0 01.696-.071l.072.071 2.5 3a.5.5 0 01-.704.704l-.064-.064L8 4.281 5.884 6.82a.5.5 0 01-.63.115l-.074-.051a.5.5 0 01-.115-.63l.051-.074 2.5-3zM7.616 12.82a.5.5 0 00.696.071l.072-.071 2.5-3a.5.5 0 00-.704-.704l-.064.064L8 11.719 5.884 9.18a.5.5 0 00-.63-.115l-.074.051a.5.5 0 00-.115.63l.051.074 2.5 3z"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rs-table-body-row-wrapper"
        role="rowgroup"
        style="top: 40px; height: 328px;"
      >
        <div
          class="rs-table-body-wheel-area"
          style="position: absolute; height: 322px; min-height: 368px;"
        >
          <div
            aria-rowindex="2"
            class="rs-table-row"
            role="row"
            style="min-width: 750px; height: 46px; transform: translate3d(0px,0px,0);"
          >
            <div
              class="rs-table-cell-group rs-table-cell-group-scroll"
              style="transform: translate3d(0px,0px,0);"
            >
              <div
                aria-colindex="1"
                class="rs-table-cell rs-table-cell-first"
                role="gridcell"
                style="width: 350px; height: 46px; z-index: 0; left: 0px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 350px; height: 46px; text-align: left;"
                >
                  <span
                    class=""
                  >
                    <a
                      href="/history/themes/motown"
                    >
                      Motown
                    </a>
                  </span>
                </div>
              </div>
              <div
                aria-colindex="2"
                class="rs-table-cell"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 350px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    20:58:47 03/05/2018
                  </span>
                </div>
              </div>
              <div
                aria-colindex="3"
                class="rs-table-cell rs-table-cell-last"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 550px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    <span
                      class="elapsed-time"
                    >
                      5 years ago
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-rowindex="3"
            class="rs-table-row"
            role="row"
            style="min-width: 750px; height: 46px; transform: translate3d(0px,46px,0);"
          >
            <div
              class="rs-table-cell-group rs-table-cell-group-scroll"
              style="transform: translate3d(0px,0px,0);"
            >
              <div
                aria-colindex="1"
                class="rs-table-cell rs-table-cell-first"
                role="gridcell"
                style="width: 350px; height: 46px; z-index: 0; left: 0px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 350px; height: 46px; text-align: left;"
                >
                  <span
                    class=""
                  >
                    <a
                      href="/history/themes/pot-luck"
                    >
                      Pot Luck
                    </a>
                  </span>
                </div>
              </div>
              <div
                aria-colindex="2"
                class="rs-table-cell"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 350px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    20:43:00 05/04/2018
                  </span>
                </div>
              </div>
              <div
                aria-colindex="3"
                class="rs-table-cell rs-table-cell-last"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 550px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    <span
                      class="elapsed-time"
                    >
                      5 years ago
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-rowindex="4"
            class="rs-table-row"
            role="row"
            style="min-width: 750px; height: 46px; transform: translate3d(0px,92px,0);"
          >
            <div
              class="rs-table-cell-group rs-table-cell-group-scroll"
              style="transform: translate3d(0px,0px,0);"
            >
              <div
                aria-colindex="1"
                class="rs-table-cell rs-table-cell-first"
                role="gridcell"
                style="width: 350px; height: 46px; z-index: 0; left: 0px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 350px; height: 46px; text-align: left;"
                >
                  <span
                    class=""
                  >
                    <a
                      href="/history/themes/pride"
                    >
                      Pride
                    </a>
                  </span>
                </div>
              </div>
              <div
                aria-colindex="2"
                class="rs-table-cell"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 350px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    20:01:00 04/06/2020
                  </span>
                </div>
              </div>
              <div
                aria-colindex="3"
                class="rs-table-cell rs-table-cell-last"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 550px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    <span
                      class="elapsed-time"
                    >
                      3 years ago
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-rowindex="5"
            class="rs-table-row"
            role="row"
            style="min-width: 750px; height: 46px; transform: translate3d(0px,138px,0);"
          >
            <div
              class="rs-table-cell-group rs-table-cell-group-scroll"
              style="transform: translate3d(0px,0px,0);"
            >
              <div
                aria-colindex="1"
                class="rs-table-cell rs-table-cell-first"
                role="gridcell"
                style="width: 350px; height: 46px; z-index: 0; left: 0px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 350px; height: 46px; text-align: left;"
                >
                  <span
                    class=""
                  >
                    <a
                      href="/history/themes/rock"
                    >
                      Rock
                    </a>
                  </span>
                </div>
              </div>
              <div
                aria-colindex="2"
                class="rs-table-cell"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 350px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    19:00:00 01/06/2020
                  </span>
                </div>
              </div>
              <div
                aria-colindex="3"
                class="rs-table-cell rs-table-cell-last"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 550px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    <span
                      class="elapsed-time"
                    >
                      3 years ago
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-rowindex="6"
            class="rs-table-row"
            role="row"
            style="min-width: 750px; height: 46px; transform: translate3d(0px,184px,0);"
          >
            <div
              class="rs-table-cell-group rs-table-cell-group-scroll"
              style="transform: translate3d(0px,0px,0);"
            >
              <div
                aria-colindex="1"
                class="rs-table-cell rs-table-cell-first"
                role="gridcell"
                style="width: 350px; height: 46px; z-index: 0; left: 0px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 350px; height: 46px; text-align: left;"
                >
                  <span
                    class=""
                  >
                    <a
                      href="/history/themes/rock-power-ballads"
                    >
                      Rock & Power Ballads
                    </a>
                  </span>
                </div>
              </div>
              <div
                aria-colindex="2"
                class="rs-table-cell"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 350px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    19:51:00 03/05/2018
                  </span>
                </div>
              </div>
              <div
                aria-colindex="3"
                class="rs-table-cell rs-table-cell-last"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 550px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    <span
                      class="elapsed-time"
                    >
                      5 years ago
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-rowindex="7"
            class="rs-table-row"
            role="row"
            style="min-width: 750px; height: 46px; transform: translate3d(0px,230px,0);"
          >
            <div
              class="rs-table-cell-group rs-table-cell-group-scroll"
              style="transform: translate3d(0px,0px,0);"
            >
              <div
                aria-colindex="1"
                class="rs-table-cell rs-table-cell-first"
                role="gridcell"
                style="width: 350px; height: 46px; z-index: 0; left: 0px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 350px; height: 46px; text-align: left;"
                >
                  <span
                    class=""
                  >
                    <a
                      href="/history/themes/sunshine"
                    >
                      sunshine
                    </a>
                  </span>
                </div>
              </div>
              <div
                aria-colindex="2"
                class="rs-table-cell"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 350px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    19:30:00 04/06/2020
                  </span>
                </div>
              </div>
              <div
                aria-colindex="3"
                class="rs-table-cell rs-table-cell-last"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 550px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    <span
                      class="elapsed-time"
                    >
                      3 years ago
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-rowindex="8"
            class="rs-table-row"
            role="row"
            style="min-width: 750px; height: 46px; transform: translate3d(0px,276px,0);"
          >
            <div
              class="rs-table-cell-group rs-table-cell-group-scroll"
              style="transform: translate3d(0px,0px,0);"
            >
              <div
                aria-colindex="1"
                class="rs-table-cell rs-table-cell-first"
                role="gridcell"
                style="width: 350px; height: 46px; z-index: 0; left: 0px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 350px; height: 46px; text-align: left;"
                >
                  <span
                    class=""
                  >
                    <a
                      href="/history/themes/various-artists"
                    >
                      Various Artists
                    </a>
                  </span>
                </div>
              </div>
              <div
                aria-colindex="2"
                class="rs-table-cell"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 350px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    15:43:56 22/04/2018
                  </span>
                </div>
              </div>
              <div
                aria-colindex="3"
                class="rs-table-cell rs-table-cell-last"
                role="gridcell"
                style="width: 200px; height: 46px; z-index: 0; left: 550px;"
              >
                <div
                  class="rs-table-cell-content"
                  style="width: 200px; height: 46px; text-align: right;"
                >
                  <span
                    class=""
                  >
                    <span
                      class="elapsed-time"
                    >
                      5 years ago
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rs-table-mouse-area"
        style="height: 368px;"
      >
        <span
          style="height: 39px;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
